File: dart/fourdgs/test/late_front_matter_test.dart
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

import 'dart:typed_data';

import 'package:fourdgs/fourdgs.dart';
import 'package:fourdgs/writer.dart';
import 'package:test/test.dart';

typedef _LateFixture =
    ({
      Uint8List bytes,
      int firstStateOpcode,
      int firstStateOffset,
      int lateOffset,
    });

const List<int> _frontMatterOpcodes = <int>[
  opHeader,
  opQuantization,
  opWindowTable,
  opAudio,
  opCamera,
  opMetadata,
  opAttachment,
  opAudioSource,
  opAudioData,
  opCoordinateFrame,
  opSensorCalibration,
  opRigTrajectory,
  opGeodeticAnchor,
  opObjectTable,
  opObjectTrack,
];

FourdgsGaussianSet _oneGaussian() => FourdgsGaussianSet(
  positions: Float32List.fromList(const <double>[0, 0, 0]),
  scales: Float32List.fromList(const <double>[1, 1, 1]),
  rotations: Float32List.fromList(const <double>[0, 0, 0, 1]),
  colors: Float32List.fromList(const <double>[0.5, 0.5, 0.5, 1]),
  motions: Float32List.fromList(const <double>[0, 0, 0]),
  muT: Float32List.fromList(const <double>[0.5]),
  sigmaT: Float32List.fromList(const <double>[0.5]),
  winLo: Float32List.fromList(const <double>[0]),
  winHi: Float32List.fromList(const <double>[1]),
);

Uint8List _gaussianBirthFile() => writeFourdgsBytes(
  _oneGaussian(),
  1.0,
  options: const FourdgsWriteOptions(
    writeIndex: false,
    writeCrc: false,
    shBands: 0,
    verify: false,
  ),
);

Uint8List _keyframeDeltaFile() => writeKeyframeDeltaBytes(
  <FourdgsSample>[
    FourdgsSample(t0: 0.0, ids: const <int>[7], gaussians: _oneGaussian()),
  ],
  1.0,
  options: const FourdgsKeyframeDeltaOptions(
    writeIndex: false,
    writeStatistics: false,
    writeCrc: false,
    verify: false,
  ),
);

Uint8List _record(int opcode, Uint8List content) {
  final header =
      ByteData(recordHeaderBytes)
        ..setUint8(0, opcode)
        ..setUint32(1, content.length, Endian.little)
        ..setUint32(5, 0, Endian.little);
  return Uint8List.fromList(<int>[...header.buffer.asUint8List(), ...content]);
}

_LateFixture _afterFirstState(
  Uint8List source,
  int opcode, {
  Uint8List? content,
}) {
  final state = iterRecords(
    source,
    fourdgsMagic.length,
  ).firstWhere((record) => isStateOpcode(record.opcode));
  final lateOffset = state.offset + recordHeaderBytes + state.content.length;
  final late = _record(opcode, content ?? Uint8List(0));
  return (
    bytes: Uint8List.fromList(<int>[
      ...source.sublist(0, lateOffset),
      ...late,
      ...source.sublist(lateOffset),
    ]),
    firstStateOpcode: state.opcode,
    firstStateOffset: state.offset,
    lateOffset: lateOffset,
  );
}

String _hex(int opcode) =>
    '0x${opcode.toRadixString(16).padLeft(2, '0').toUpperCase()}';

Matcher _diagnostic(_LateFixture fixture, int lateOpcode) => allOf(
  contains(
    '${opcodeName(lateOpcode)} record (opcode ${_hex(lateOpcode)}) '
    'at byte ${fixture.lateOffset}',
  ),
  contains(
    'first state record, ${opcodeName(fixture.firstStateOpcode)} '
    '(opcode ${_hex(fixture.firstStateOpcode)}) at byte '
    '${fixture.firstStateOffset}',
  ),
  contains('expected every defined front-matter record before'),
);

Matcher _lateRefusal(_LateFixture fixture, int lateOpcode) =>
    isA<FourdgsMalformedFile>()
        .having(
          (error) => error.refusalCode,
          'refusal code',
          refusalLateFrontMatterRecord,
        )
        .having(
          (error) => error.message,
          'physical record diagnosis',
          _diagnostic(fixture, lateOpcode),
        );

Future<void> _expectValidatorRefusal(
  _LateFixture fixture,
  int lateOpcode,
) async {
  final report = await validateFourdgs(FourdgsBytes(fixture.bytes));
  final refused = report.findings.where(
    (finding) => finding.refusal?.code == refusalLateFrontMatterRecord,
  );
  expect(refused, hasLength(1));
  final finding = refused.single;
  expect(finding.message, _diagnostic(fixture, lateOpcode));
  expect(finding.refusal!.site!.offset, fixture.lateOffset);
  expect(finding.refusal!.site!.what, 'the ${opcodeName(lateOpcode)} record');
}

void main() {
  group('late-front-matter-record', () {
    final gaussianBirth = _gaussianBirthFile();
    final keyframeDelta = _keyframeDeltaFile();

    for (final opcode in _frontMatterOpcodes) {
      test(
        '${opcodeName(opcode)} is refused before duplicate or body parsing',
        () async {
          // Empty bodies make every structured member hostile. Header,
          // Quantization, and WindowTable are also late duplicates. Reaching a
          // body parser or multiplicity check would therefore produce the wrong
          // refusal, so this one table pins both precedence requirements.
          final gaussianFixture = _afterFirstState(gaussianBirth, opcode);
          final deltaFixture = _afterFirstState(keyframeDelta, opcode);

          expect(
            () => readFourdgsBytes(
              gaussianFixture.bytes,
              recoverTruncated: false,
            ),
            throwsA(_lateRefusal(gaussianFixture, opcode)),
          );
          expect(
            () => decodeKeyframeDeltaStreamed(deltaFixture.bytes),
            throwsA(_lateRefusal(deltaFixture, opcode)),
          );
          await _expectValidatorRefusal(gaussianFixture, opcode);
        },
      );
    }

    test('the first state record may itself be a Delta Chunk', () async {
      final source = _gaussianBirthFile();
      final state = iterRecords(
        source,
        fourdgsMagic.length,
      ).firstWhere((record) => record.opcode == opChunk);
      final withDeltaOpcode = Uint8List.fromList(source)
        ..[state.offset] = opDeltaChunk;
      final fixture = _afterFirstState(withDeltaOpcode, opHeader);

      expect(
        () => readFourdgsBytes(fixture.bytes, recoverTruncated: false),
        throwsA(_lateRefusal(fixture, opHeader)),
      );
      await _expectValidatorRefusal(fixture, opHeader);
    });

    for (final opcode in <int>[0x7F, 0x80]) {
      test('${opcodeName(opcode)} remains position-independent', () async {
        expect(isFrontMatterOpcode(opcode), isFalse);
        final gaussianFixture = _afterFirstState(gaussianBirth, opcode);
        final deltaFixture = _afterFirstState(keyframeDelta, opcode);

        final scene = readFourdgsBytes(
          gaussianFixture.bytes,
          recoverTruncated: false,
        );
        expect(scene.gaussians.count, 1);
        expect(scene.skippedOpcodes, contains(opcode));
        expect(
          decodeKeyframeDeltaStreamed(deltaFixture.bytes).chunks,
          hasLength(1),
        );

        final report = await validateFourdgs(
          FourdgsBytes(gaussianFixture.bytes),
        );
        expect(report.ok, isTrue);
        expect(
          report.findings.where(
            (finding) => finding.refusal?.code == refusalLateFrontMatterRecord,
          ),
          isEmpty,
        );
      });
    }

    test('the placement classifier is the registry closed set', () {
      expect(<int>[
        for (int opcode = 0; opcode <= 0xFF; opcode++)
          if (isFrontMatterOpcode(opcode)) opcode,
      ], _frontMatterOpcodes);
      expect(isStateOpcode(opChunk), isTrue);
      expect(isStateOpcode(opDeltaChunk), isTrue);
      expect(isStateOpcode(0x26), isFalse, reason: 'reserved stays unassigned');
    });
  });
}

File: typescript/conformance/src/lateFrontMatter.test.ts
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

import assert from "node:assert/strict";
import { test } from "node:test";

import {
  BytesReadable,
  IndexedDecoder,
  MAGIC,
  MalformedFile,
  Opcode,
  Refusal,
  decodeKeyframeDeltaStreamed,
  decodeScene,
  encodeKeyframeDeltaSequence,
  encodeScene,
  isFrontMatterOpcode,
  isStateOpcode,
  iterateRecords,
  keyframeDeltaValidationRecordOffset,
  opcodeName,
  validateKeyframeDeltaStreamed,
  type GaussianInput,
} from "@4dgs/core";
import { validateFile } from "@4dgs/nodejs";

import { concat, record } from "./testing.js";

interface LateFixture {
  readonly bytes: Uint8Array;
  readonly firstStateOpcode: number;
  readonly firstStateOffset: number;
  readonly lateOffset: number;
}

const FRONT_MATTER_OPCODES: readonly number[] = [
  Opcode.Header,
  Opcode.Quantization,
  Opcode.WindowTable,
  Opcode.Audio,
  Opcode.Camera,
  Opcode.Metadata,
  Opcode.Attachment,
  Opcode.AudioSource,
  Opcode.AudioData,
  Opcode.CoordinateFrame,
  Opcode.SensorCalibration,
  Opcode.RigTrajectory,
  Opcode.GeodeticAnchor,
  Opcode.ObjectTable,
  Opcode.ObjectTrack,
];

const gaussian: GaussianInput = {
  count: 1,
  positions: [0, 0, 0],
  scales: [1, 1, 1],
  rotations: [0, 0, 0, 1],
  colors: [0.5, 0.5, 0.5, 1],
  motions: [0, 0, 0],
  muT: [0.5],
  sigmaT: [0.5],
  winLo: [0],
  winHi: [1],
};

const gaussianBirth = encodeScene(gaussian, 1, {
  writeIndex: false,
  writeCrc: false,
  shBands: 0,
});
const keyframeDelta = encodeKeyframeDeltaSequence([{ t0: 0, ids: [7], gaussians: gaussian }], 1, {
  writeIndex: false,
  writeStatistics: false,
  writeCrc: false,
});

function afterFirstState(source: Uint8Array, opcode: number): LateFixture {
  const state = [...iterateRecords(source, MAGIC.length)].find((item) =>
    isStateOpcode(item.opcode),
  );
  assert.ok(state !== undefined, "the writer emitted a state record");
  const lateOffset = state.offset + state.length;
  return {
    bytes: concat([
      source.subarray(0, lateOffset),
      record(opcode, new Uint8Array(0)),
      source.subarray(lateOffset),
    ]),
    firstStateOpcode: state.opcode,
    firstStateOffset: state.offset,
    lateOffset,
  };
}

function hex(opcode: number): string {
  return `0x${opcode.toString(16).padStart(2, "0").toUpperCase()}`;
}

function checkDiagnosis(error: unknown, fixture: LateFixture, lateOpcode: number): boolean {
  assert.ok(error instanceof MalformedFile, String(error));
  assert.equal(error.refusalCode, Refusal.LateFrontMatterRecord);
  assert.match(
    error.message,
    new RegExp(
      `${opcodeName(lateOpcode)} record \\(opcode ${hex(lateOpcode)}\\) at byte ` +
        `${fixture.lateOffset}`,
    ),
  );
  assert.match(
    error.message,
    new RegExp(
      `first state record, ${opcodeName(fixture.firstStateOpcode)} \\(opcode ` +
        `${hex(fixture.firstStateOpcode)}\\) at byte ${fixture.firstStateOffset}`,
    ),
  );
  assert.match(error.message, /expected every defined front-matter record before/);
  return true;
}

async function checkValidator(fixture: LateFixture, lateOpcode: number): Promise<void> {
  const report = await validateFile(new BytesReadable(fixture.bytes));
  assert.equal(report.ok, false);
  assert.equal(report.refused?.code, Refusal.LateFrontMatterRecord);
  assert.equal(report.refused?.at, fixture.lateOffset);
  assert.equal(report.refused?.where, `the ${opcodeName(lateOpcode)} record`);
  checkDiagnosis(
    new MalformedFile(report.refused!.message, { refusalCode: report.refused!.code }),
    fixture,
    lateOpcode,
  );
}

for (const opcode of FRONT_MATTER_OPCODES) {
  test(`${opcodeName(opcode)} is refused before duplicate or body parsing`, async () => {
    // Every inserted body is empty and therefore hostile to its structured parser.
    // Header, Quantization and WindowTable are also duplicates. Only a placement
    // check before both semantics can give all fifteen records the same answer.
    const gaussianFixture = afterFirstState(await gaussianBirth, opcode);
    const deltaFixture = afterFirstState(await keyframeDelta, opcode);

    await assert.rejects(
      () => decodeScene(new BytesReadable(gaussianFixture.bytes), { recoverTruncated: false }),
      (error: unknown) => checkDiagnosis(error, gaussianFixture, opcode),
    );
    await assert.rejects(
      () => decodeKeyframeDeltaStreamed(deltaFixture.bytes),
      (error: unknown) => checkDiagnosis(error, deltaFixture, opcode),
    );
    await checkValidator(gaussianFixture, opcode);
    await assert.rejects(
      () => validateKeyframeDeltaStreamed(deltaFixture.bytes),
      (error: unknown) => {
        assert.equal(keyframeDeltaValidationRecordOffset(error), deltaFixture.lateOffset);
        return checkDiagnosis(error, deltaFixture, opcode);
      },
    );
  });
}

test("the first state record may itself be a DeltaChunk", async () => {
  const source = Uint8Array.from(await gaussianBirth);
  const state = [...iterateRecords(source, MAGIC.length)].find(
    (item) => item.opcode === Opcode.Chunk,
  );
  assert.ok(state !== undefined);
  source[state.offset] = Opcode.DeltaChunk;
  const fixture = afterFirstState(source, Opcode.Header);

  await assert.rejects(
    () => decodeScene(fixture.bytes, { recoverTruncated: false }),
    (error: unknown) => checkDiagnosis(error, fixture, Opcode.Header),
  );
  await checkValidator(fixture, Opcode.Header);
});

for (const opcode of [0x7f, 0x80]) {
  test(`${opcodeName(opcode)} remains position-independent`, async () => {
    assert.equal(isFrontMatterOpcode(opcode), false);
    const gaussianFixture = afterFirstState(await gaussianBirth, opcode);
    const deltaFixture = afterFirstState(await keyframeDelta, opcode);

    const scene = await decodeScene(gaussianFixture.bytes, { recoverTruncated: false });
    assert.equal(scene.gaussians.count, 1);
    assert.ok(scene.skippedOpcodes.includes(opcode));
    assert.equal((await decodeKeyframeDeltaStreamed(deltaFixture.bytes)).chunks.length, 1);
    assert.equal(await validateKeyframeDeltaStreamed(deltaFixture.bytes), 1);

    const report = await validateFile(gaussianFixture.bytes);
    assert.equal(report.ok, true, report.findings.map((finding) => finding.message).join("\n"));
    assert.equal(report.refused, null);
  });
}

test("the placement classifier is the registry's closed set", () => {
  assert.deepEqual(
    Array.from({ length: 256 }, (_, opcode) => opcode).filter(isFrontMatterOpcode),
    FRONT_MATTER_OPCODES,
  );
  assert.equal(isStateOpcode(Opcode.Chunk), true);
  assert.equal(isStateOpcode(Opcode.DeltaChunk), true);
  assert.equal(isStateOpcode(0x26), false, "reserved opcodes stay unassigned");
});

test("indexed open stops at state; deferred discovery names the late record", async () => {
  const fixture = afterFirstState(await gaussianBirth, Opcode.Metadata);
  const reads: { readonly offset: number; readonly length: number }[] = [];
  const source = {
    size: () => Promise.resolve(BigInt(fixture.bytes.byteLength)),
    read: (offset: bigint, length: bigint): Promise<Uint8Array> => {
      const at = Number(offset);
      const count = Number(length);
      reads.push({ offset: at, length: count });
      return Promise.resolve(fixture.bytes.subarray(at, at + count));
    },
  };
  const opened = await IndexedDecoder.open(source, { headProbeBytes: 32 });
  assert.ok(
    reads.every(
      ({ offset, length }) => fixture.lateOffset < offset || fixture.lateOffset >= offset + length,
    ),
    "indexed open must not scan across state to discover the late record",
  );
  await assert.rejects(
    () => opened.readMetadata(),
    (error: unknown) => checkDiagnosis(error, fixture, Opcode.Metadata),
  );
});
